RecFCCeeCalorimeter/src/components/PhotonIDTool.cpp

```diff
@@ -0,0 +1,352 @@
+	}
+      else {
+	m_inputPositionsInShapeParameters.push_back(-1);
+      }
+    }
+    else {
+      // for the other features, check if they are in the shape paramaters
+      auto it = std::find(shapeParameters.begin(), shapeParameters.end(), feature);
+      if (it != shapeParameters.end())
+	{
+	  int position = std::distance(shapeParameters.begin(), it);
+	  m_inputPositionsInShapeParameters.push_back(position);
+	  info() << "Feature " << feature << " found in position " << position << " of shapeParameters" << endmsg;
+	}
+      else
+	{
+	  // at least one of the inputs of the MVA was not found in the shapeParameters
+	  // so we can stop checking the others 
+	  m_inputPositionsInShapeParameters.clear();
+	  error() << "Feature " << feature << " not found, aborting..." << endmsg;
+	  return StatusCode::FAILURE;
+	}
+    }
+  }
+    
+  // append the MVA score to the output shape parameters
+  shapeParameters.push_back("photonIDscore");
+  m_outShapeParameterHandle.put(shapeParameters);
+
+  info() << "Initialized the photonID MVA tool" << endmsg;
+  return StatusCode::SUCCESS;
+}
+
+StatusCode PhotonIDTool::execute([[maybe_unused]] const EventContext &evtCtx) const
+{
+  verbose() << "-------------------------------------------" << endmsg;
+
+  // Get the input collection with clusters
+  const edm4hep::ClusterCollection *inClusters = m_inClusters.get();
+
+  // Initialize output clusters
+  edm4hep::ClusterCollection *outClusters = initializeOutputClusters(inClusters);
+  if (!outClusters)
+  {
+    error() << "Something went wrong in initialization of the output cluster collection, exiting!" << endmsg;
+    return StatusCode::FAILURE;
+  }
+  if (inClusters->size() != outClusters->size())
+  {
+    error() << "Sizes of input and output cluster collections does not match, exiting!" << endmsg;
+    return StatusCode::FAILURE;
+  }
+
+  // Run inference
+  {
+    StatusCode sc = applyMVAtoClusters(inClusters, outClusters);
+    if (sc.isFailure())
+    {
+      return sc;
+    }
+  }
+
+  return StatusCode::SUCCESS;
+}
+
+StatusCode PhotonIDTool::finalize()
+{
+  if (m_ortSession)
+    delete m_ortSession;
+  if (m_ortEnv)
+    delete m_ortEnv;
+
+  return Gaudi::Algorithm::finalize();
+}
+
+edm4hep::ClusterCollection *PhotonIDTool::initializeOutputClusters(
+    const edm4hep::ClusterCollection *inClusters) const
+{
+  edm4hep::ClusterCollection *outClusters = m_outClusters.createAndPut();
+
+  for (auto const &inCluster : *inClusters)
+  {
+    auto outCluster = inCluster.clone();
+    outClusters->push_back(outCluster);
+  }
+
+  return outClusters;
+}
+
+StatusCode PhotonIDTool::readMVAFiles(const std::string& mvaInputsFileName,
+				      const std::string& mvaModelFileName)
+{
+  // 1. read the file with the list of input features
+  // TODO: figure how to save BDT to ONNX such that input is
+  // the list of input features with their names, rather
+  // than single tensor "X"
```

Comment:
Can you use `nlohmann_json` instead of parsing the JSON manually? It is a dependency anyway (make it explicit in the CMake file). To keep ability to have comments:
https://json.nlohmann.me/features/comments/

Also, why are you splitting part of the configuration outside of the steering file?
Do you have that many shape parameters?

If this is just temporary solution, ignore this comment.

Reply:
Hi @kjvbrt, the number of shape parameters can exceed 100.
What are you referring to in particular when you write "splitting part of the configuration outside of the steering file"? Do you mean, putting the names of the shape variables in the json rather than in the .py that steers k4run? Or something else?
About nlohmann_json, I thought it was an overkill for just parsing a json that contains a list of strings (could have been a CSV). But OK, in the end I can probably use that, so that in the json I will also save the parameters used at training time which is always useful to record somewhere

Reply:
> What are you referring to in particular when you write "splitting part of the configuration outside of the steering file"? Do you mean, putting the names of the shape variables in the json rather than in the .py that steers k4run? Or something else?

Yes, this is what I mean, I would like to keep all configurations needed for running in one file, if possible. I agree that 100 parameters is a bit too much for the k4run steering file :)

> About nlohmann_json, I thought it was an overkill for just parsing a json that contains a list of strings (could have been a CSV). But OK, in the end I can probably use that, so that in the json I will also save the parameters used at training time which is always useful to record somewhere

I would not suggest using it if it was not already a hidden dependency.

What is your TODO about, is it that you might be able read the list of the shape parameters directly from the ONNX file?

Reply:
yes, the TODO was about that, but it seems not possible unfortunately

Reply:
Hi @kjvbrt I implemented the parsing of the json with nlohmann. The json stores all relevant info from the training of the photonID algorithm and is generated automatically by the tool that trains the BDT and creates the onnx model, so using it rather than copying to the k4run steering file in my opinion provides some better consistency.
I saw that you merged some PRs so I will probably have to rebase my PR. Apart from that is there anything else I should change? Thanks a lot!